### PR TITLE
.github/workflows/snap-build.yaml: build the snapd snap via GH Actions too

### PR DESCRIPTION
**Pull request by anonymouse64**
This will be useful if folks want to test a build of snapd from a branch or from a PR, as the artifact will be built and available for testing by just downloading it from the Actions page.

### CODE REVIEW

File: .github/workflows/snap-build.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        uses: actions/checkout@v2
+      - name: Build the snapd snap
+        uses: snapcore/action-build@v1
+      - name: Rename built snap file
+        run: |
+          sudo mv snapd*.snap snapd.snap
+      - name: Uploading snap artifact
+        uses: actions/upload-artifact@v1
+        with:
+          name: snapd.snap
```
Comment (zyga) [inline .github/workflows/snap-build.yaml:20]: Could you have used a glob here? I'm not even sure what's here but I think preserving the architecture would be pretty useful.

**Comment by bboozzoo** [comment]
How long are the artifacts preserved?

File: .github/workflows/snap-build.yaml (hunk 1 above)
Comment (anonymouse64) [reply .github/workflows/snap-build.yaml:20]: Mmm I didn't try a glob, I can see what happens with a glob easily enough

Comment (anonymouse64) [reply .github/workflows/snap-build.yaml:20]: no globs do not work unfortunately, see my etrace test run:

```
2020-07-22T18:47:20.3406383Z Snapped etrace_2a130c0_amd64.snap
2020-07-22T18:47:21.7326601Z ##[group]Run actions/upload-artifact@v1
2020-07-22T18:47:21.7326814Z with:
2020-07-22T18:47:21.7326966Z   name: etrace*.snap
2020-07-22T18:47:21.7327163Z   path: etrace.snap
2020-07-22T18:47:21.7327333Z ##[endgroup]
2020-07-22T18:47:21.9823285Z ##[error]Path does not exist /home/runner/work/etrace/etrace/etrace.snap
2020-07-22T18:47:22.0136894Z ##[error]Exit code 1 returned from process: file name '/home/runner/runners/2.267.1/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'.
2020-07-22T18:47:22.0191116Z Post job cleanup.
```

**Comment by anonymouse64** [comment]
> How long are the artifacts preserved?

from https://docs.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts#about-workflow-artifacts, it seems 90 days, unless a new push is done to a PR in which case the old one is dropped it seems:

> For example, you can use artifacts to save your build and test output after a workflow run has ended. For pushes and pull requests, GitHub stores artifacts for 90 days. The retention period for a pull request restarts each time someone pushes a new commit to the pull request.

**Comment by anonymouse64** [comment]
> Could the re-executing systems pull the file instead of building it from scratch each time?

Actually yes, it seems they can, I see there is a corresponding action, https://github.com/actions/download-artifact for dowloading artifacts. Probably for snapd, I don't think we actually consume the snapd snap as built anywhere in the spread tests, but I think that if we have actions to build the various debs, rpms, etc. those could be consumed with the download-artifact action.

File: .github/workflows/snap-build.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Build the snapd snap
+        uses: snapcore/action-build@v1
+      - name: Rename built snap file
+        run: |
+          sudo mv snapd*.snap snapd.snap
+      - name: Uploading snap artifact
+        uses: actions/upload-artifact@v1
```
Comment (cmatsuoka) [inline .github/workflows/snap-build.yaml:18]: Any special reason to use v1 instead of v2?

File: .github/workflows/snap-build.yaml (hunk 1 above)
Comment (cmatsuoka) [reply .github/workflows/snap-build.yaml:20]: Maybe it works in v2? Globbing is mentioned in the docs.

Comment (anonymouse64) [reply .github/workflows/snap-build.yaml:20]: ah interesting I didn't even realize I was using v2 instead of v1, let me try again with v2 and globbing...

File: .github/workflows/snap-build.yaml (hunk 2 above)
Comment (anonymouse64) [reply .github/workflows/snap-build.yaml:18]: that was just what the example I was using had :smiley:

Comment (anonymouse64) [reply .github/workflows/snap-build.yaml]: v2 works the same but also allows globbing, I was using it wrong

File: .github/workflows/snap-build.yaml (hunk 1 above)
Comment (anonymouse64) [reply .github/workflows/snap-build.yaml]: it does work :tada:

**Comment by jhenstridge** [comment]
> Actually yes, it seems they can, I see there is a corresponding action, https://github.com/actions/download-artifact for dowloading artifacts. Probably for snapd, I don't think we actually consume the snapd snap as built anywhere in the spread tests, but I think that if we have actions to build the various debs, rpms, etc. those could be consumed with the download-artifact action.

If you want to use `upload-artifact`/`download-artifact` to transfer data between jobs.  Artifacts are per-workflow, and you want to make sure you have an ordering dependency between the two jobs so the download happens after the upload.

That would push towards adding this to the big `test.yaml` workflow.

File: .github/workflows/snap-build.yaml (hunk 1 above)
Comment (jhenstridge) [reply .github/workflows/snap-build.yaml]: fwiw, `snapcore/action-build` makes the snap file name available as an output parameter.  That was more important back before `upload-artifact@v2` was released.

**Comment by anonymouse64** [comment]
> That would push towards adding this to the big test.yaml workflow.

I think considering that this would take some time on our part to re-engineer how the spread tests run that is a non-trivial amount of time, I would rather leave this here as another workflow, as that will run independently of the "Tests" workflow and thus won't add any time to that workflow. When the time comes to try and do this, it's easy enough to move these steps to the big "Tests" workflow, but we can benefit right now from having this additional workflow that just uploads the snapd snap built from a branch.
